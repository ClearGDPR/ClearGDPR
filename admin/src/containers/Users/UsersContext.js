import React, { Component, createContext } from 'react';
import PropTypes from 'prop-types';
import config from 'config';
import internalFetch from 'helpers/internal-fetch';

const UsersContext = createContext({
  users: [],
  fetchUsers: () => {},
  registerUser: () => {},
  deleteUser: () => {},
  isLoading: false
});

export class UsersProvider extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.element)
    ])
  };

  state = {
    users: [],
    fetchUsers: this.fetchUsers.bind(this),
    registerUser: this.registerUser.bind(this),
    deleteUser: this.deleteUser.bind(this),
    isLoading: false
  };

  _getUsers() {
    return internalFetch(`${config.API_URL}/api/management/users/list`);
  }

  _registerUser(username, password) {
    return internalFetch(`${config.API_URL}/api/management/users/register`, {
      method: 'POST',
      body: JSON.stringify({
        username,
        password
      })
    });
  }

  async _deleteUser(userId) {
    await internalFetch(`${config.API_URL}/api/management/users/${userId}/remove`, {
      method: 'POST'
    });
  }

  setLoading(loading) {
    this.setState({
      isLoading: loading
    });
  }

  cancelLoadingAndReject(e) {
    this.setLoading(false);
    return Promise.reject(e);
  }

  async fetchUsers() {
    this.setLoading(true);

    let users = await this._getUsers();
    this.setState({
      users,
      isLoading: false
    });
  }

  async registerUser(username, password) {
    this.setLoading(true);

    const user = await this._registerUser(username, password).catch(
      this.cancelLoadingAndReject.bind(this)
    );
    await this.fetchUsers();

    return user;
  }

  async deleteUser(userId) {
    this.setLoading(true);
    await this._deleteUser(userId).catch(this.cancelLoadingAndReject.bind(this));
    await this.fetchUsers();
  }

  render() {
    return <UsersContext.Provider value={this.state}>{this.props.children}</UsersContext.Provider>;
  }
}

export const UsersConsumer = UsersContext.Consumer;
