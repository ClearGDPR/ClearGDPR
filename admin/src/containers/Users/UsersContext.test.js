import React from 'react';

import { shallow } from 'enzyme';

import { UsersProvider } from './UsersContext';
import session from '../../helpers/Session';
import TestUtils from '../../helpers/TestUtils';

jest.mock('../../helpers/Session');

beforeEach(() => {
  session.getToken.mockReturnValue('token');
});

describe('UsersProvider', () => {
  let users = [
    {
      id: 1,
      username: 'admin'
    },
    {
      id: 2,
      username: 'joe'
    },
    {
      id: 3,
      username: 'marc'
    },
    {
      id: 4,
      username: 'gina'
    }
  ];

  const setupShallow = propOverrides => {
    const props = Object.assign({}, propOverrides);
    const component = shallow(<UsersProvider {...props} />);

    return { props, component };
  };

  it('should have correct state after fetching users', async () => {
    global.fetch = jest.fn().mockImplementationOnce(() =>
      Promise.resolve({
        status: 200,
        json: () => {
          return users;
        }
      })
    );

    const { component } = setupShallow();
    component.instance().fetchUsers();

    await TestUtils.flushPromises();

    expect(component.state()).toEqual(
      expect.objectContaining({
        users: users,
        isLoading: false
      })
    );
  });

  it('should throw error when registering new user returned HTTP 400', async () => {
    global.fetch = jest.fn().mockImplementationOnce(() =>
      Promise.resolve({
        status: 400,
        json: () => ({ error: 'Error message' })
      })
    );

    const { component } = setupShallow();

    await expect(
      component.instance().registerUser('bobby', 'somePassword123')
    ).rejects.toMatchSnapshot();
    expect(component.state()).toEqual(
      expect.objectContaining({
        isLoading: false
      })
    );
  });

  it('should throw error when registering new user returned HTTP 500', async () => {
    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        status: 500
      })
    );

    const { component } = setupShallow();

    await expect(
      component.instance().registerUser('bobby', 'somePassword123')
    ).rejects.toMatchSnapshot();
    expect(component.state()).toEqual(
      expect.objectContaining({
        isLoading: false
      })
    );
  });

  it('should have correct state after registering a new user', async () => {
    let i = 0;
    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        status: 201,
        json: () => {
          const result =
            i === 0 ? { id: 5, username: 'bobby' } : [...users, { id: 5, username: 'bobby' }];
          i++;
          return result;
        }
      })
    );

    const { component } = setupShallow();
    const bobby = await component.instance().registerUser('bobby', 'somePassword123');

    await TestUtils.flushPromises();

    expect(component.state()).toEqual(
      expect.objectContaining({
        users: expect.arrayContaining([...users, bobby]),
        isLoading: false
      })
    );
  });

  it('should render correctly', async () => {
    global.fetch = jest.fn().mockImplementationOnce(() =>
      Promise.resolve({
        status: 200,
        json: () => {
          return users;
        }
      })
    );

    const { component } = setupShallow();

    await TestUtils.flushPromises();
    expect(component).toMatchSnapshot();
  });
});
