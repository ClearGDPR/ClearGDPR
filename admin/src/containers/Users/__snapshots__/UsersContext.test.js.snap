// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersProvider should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UsersProvider />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": undefined,
      "value": Object {
        "deleteUser": [Function],
        "fetchUsers": [Function],
        "isLoading": false,
        "registerUser": [Function],
        "users": Array [],
      },
    },
    "ref": null,
    "rendered": null,
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_changedBits2": 0,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "deleteUser": [Function],
          "fetchUsers": [Function],
          "isLoading": false,
          "registerUser": [Function],
          "users": Array [],
        },
        "_currentValue2": Object {
          "deleteUser": [Function],
          "fetchUsers": [Function],
          "isLoading": false,
          "registerUser": [Function],
          "users": Array [],
        },
        "_defaultValue": Object {
          "deleteUser": [Function],
          "fetchUsers": [Function],
          "isLoading": false,
          "registerUser": [Function],
          "users": Array [],
        },
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": undefined,
        "value": Object {
          "deleteUser": [Function],
          "fetchUsers": [Function],
          "isLoading": false,
          "registerUser": [Function],
          "users": Array [],
        },
      },
      "ref": null,
      "rendered": null,
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {
            "deleteUser": [Function],
            "fetchUsers": [Function],
            "isLoading": false,
            "registerUser": [Function],
            "users": Array [],
          },
          "_currentValue2": Object {
            "deleteUser": [Function],
            "fetchUsers": [Function],
            "isLoading": false,
            "registerUser": [Function],
            "users": Array [],
          },
          "_defaultValue": Object {
            "deleteUser": [Function],
            "fetchUsers": [Function],
            "isLoading": false,
            "registerUser": [Function],
            "users": Array [],
          },
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`UsersProvider should throw error when deleting a user returned HTTP 404 1`] = `[Error: Error message]`;

exports[`UsersProvider should throw error when deleting a user returned HTTP 500 1`] = `[Error: Unknown error occurred]`;

exports[`UsersProvider should throw error when registering new user returned HTTP 400 1`] = `[Error: Error message]`;

exports[`UsersProvider should throw error when registering new user returned HTTP 500 1`] = `[Error: Unknown error occurred]`;
